colosseum/parser.py

```diff
@@ -179,3 +180,81 @@ def rect(value):
             return Rect(*values)
 
     raise ValueError('Unknown shape %s' % value)
+
+
+def uri(value):
+    """Parse a url from a value.
+
+    Accepts:
+    * url("<url>")
+    * url(  "<url>"  )
+    * url('<url>')
+    * url(  '<url>'  )
+    * url(<url>)
+    * url(  <url>  )
+    """
+    if isinstance(value, str):
+        value = value.strip()
+    else:
+        raise ValueError('Value {value} must be a string')
+
+    if value.startswith('url(') and value.endswith(')'):
+        # Remove the 'url(' and ')'
+        value = value[4:-1].strip()
+
+        # Single quotes and optional spaces
+        if value.startswith("'") and value.endswith("'"):
+            return Uri(value[1:-1])
+        # Double quotes and optional spaces
+        elif value.startswith('"') and value.endswith('"'):
+            # '("some.url")'
+            return Uri(value[1:-1])
+        # No quotes and optional spaces
+        else:
+            # Some characters appearing in an unquoted URI, such as parentheses, white space characters,
+            # single quotes (') and double quotes ("), must be escaped with a backslash so that the
+            # resulting URI value is a URI token: '\(', '\)'
+            escape_chars = ['(', ')', ' ', "'", '"']
+            for char in escape_chars:
+                if char in value and r'\{char}'.format(char=char) not in value:
```

Comment:
This won't be as robust as we need it to be. `"this ( is a \( string"` will pass this check. You need to do a complete check that all "special" characters are escaped.

Reply:
Also - do we actually need to check `()`? And check for `'` and `"` escaping in all cases? We only need to escape `"` if the original string was enclosed in `"` - if it's a `'` string, an `"` is unambiguous.

Reply:
>This won't be as robust as we need it to be. "this ( is a \( string" will pass this check. You need to do a complete check that all "special" characters are escaped.

Yes, I will fix this!.

Reply:
So according to https://developer.mozilla.org/en-US/docs/Web/CSS/url():

>A url, which is a relative or absolute address, or pointer, to the web resource to be included, or a data uri, optionally in single or double quotes. Quotes are required if the URL includes parentheses, whitespace, or quotes, unless these characters are escaped, or if the address includes control characters above 0x7ez .  Double quotes cannot occur inside double quotes and single quotes cannot occur inside single quotes unless escaped. 

Not sure what  `or if the address includes control characters above 0x7ez` means

>Double quotes cannot occur inside double quotes and single quotes cannot occur inside single quotes unless escaped. 

I can add a check for this then

Regarding:
>Quotes are required if the URL includes parentheses, whitespace, or quotes, unless these characters are escaped, or if the address includes control characters above 0x7ez . 

It talks about parens, have not found an example but they need to be between quotes or otehrwise escaped.